### ISSUE TITLE
unable to spin up fourth node after rolling upgrade

### ISSUE THREAD

**Issue by dianasaur323** [cockroachdb/cockroach#21516]
I had a functioning four node cluster running on DO. It has localities set, but all of the nodes are located in NY zones. I upgraded to the most recent alpha, and have been unable to spin up my fourth node. Replica rebalancing has stopped and is unable to move replicas over to the three remaining live nodes.

Here are logs from the node that keeps dying: 
```I180117 17:46:21.042444 137 storage/store.go:4291  [n4,s4] temporarily disabling rebalances because RaftLogSelfBehindCount=229229
I180117 17:46:21.645308 2349 storage/replica_raftstorage.go:746  [n4,s4,r867/3:/Table/66/1/"{fgK\xa…-nzm\x1…}] applying Raft snapshot at index 565 (id=99046079, encoded size=42697772, 163 rocksdb batches, 8 log entries)
I180117 17:46:22.470859 2349 storage/replica_raftstorage.go:752  [n4,s4,r867/3:/Table/66/1/"{fgK\xa…-nzm\x1…}] applied Raft snapshot in 825ms [clear=0ms batch=499ms entries=0ms commit=325ms]
W180117 17:46:24.100418 179 storage/node_liveness.go:426  [n4,hb] slow heartbeat took 4.5s
W180117 17:46:24.100650 179 storage/node_liveness.go:365  [n4,hb] failed node liveness heartbeat: context deadline exceeded
W180117 17:46:28.600948 179 storage/node_liveness.go:426  [n4,hb] slow heartbeat took 4.5s
W180117 17:46:28.601117 179 storage/node_liveness.go:365  [n4,hb] failed node liveness heartbeat: context deadline exceeded
I180117 17:46:31.025877 140 server/status/runtime.go:219  [n4] runtime stats: 693 MiB RSS, 170 goroutines, 234 MiB/297 MiB/566 MiB GO alloc/idle/total, 122 MiB/144 MiB CGO alloc/total, 281.51cgo/sec, 1.18/0.06 %(u/s)time, 0.00 %gc (2x)
I180117 17:46:31.044009 137 storage/store.go:4291  [n4,s4] temporarily disabling rebalances because RaftLogSelfBehindCount=228228
W180117 17:46:33.101389 179 storage/node_liveness.go:426  [n4,hb] slow heartbeat took 4.5s
W180117 17:46:33.101576 179 storage/node_liveness.go:365  [n4,hb] failed node liveness heartbeat: context deadline exceeded
W180117 17:46:37.601852 179 storage/node_liveness.go:426  [n4,hb] slow heartbeat took 4.5s
W180117 17:46:37.601994 179 storage/node_liveness.go:365  [n4,hb] failed node liveness heartbeat: context deadline exceeded
I180117 17:46:41.028936 140 server/status/runtime.go:219  [n4] runtime stats: 696 MiB RSS, 170 goroutines, 256 MiB/276 MiB/566 MiB GO alloc/idle/total, 124 MiB/147 MiB CGO alloc/total, 280.31cgo/sec, 0.98/0.04 %(u/s)time, 0.00 %gc (0x)
I180117 17:46:41.037951 137 storage/store.go:4291  [n4,s4] temporarily disabling rebalances because RaftLogSelfBehindCount=228228
W180117 17:46:42.102320 179 storage/node_liveness.go:426  [n4,hb] slow heartbeat took 4.5s
W180117 17:46:42.102549 179 storage/node_liveness.go:365  [n4,hb] failed node liveness heartbeat: context deadline exceeded
W180117 17:46:46.602920 179 storage/node_liveness.go:426  [n4,hb] slow heartbeat took 4.5s
W180117 17:46:46.603152 179 storage/node_liveness.go:365  [n4,hb] failed node liveness heartbeat: context deadline exceeded
I180117 17:46:51.023812 138 gossip/gossip.go:487  [n4] gossip status (ok, 4 nodes)
gossip client (1/3 cur/max conns)
  1: 159.203.99.113:26257 (9m1s: infos 14/3310 sent/received, bytes 4450B/466634B sent/received)
gossip server (0/3 cur/max conns, infos 14/3310 sent/received, bytes 4450B/466634B sent/received)
I180117 17:46:51.026263 140 server/status/runtime.go:219  [n4] runtime stats: 699 MiB RSS, 170 goroutines, 274 MiB/260 MiB/566 MiB GO alloc/idle/total, 127 MiB/151 MiB CGO alloc/total, 258.07cgo/sec, 0.98/0.04 %(u/s)time, 0.00 %gc (0x)
I180117 17:46:51.043565 137 storage/store.go:4291  [n4,s4] temporarily disabling rebalances because RaftLogSelfBehindCount=228228
W180117 17:46:51.103500 179 storage/node_liveness.go:426  [n4,hb] slow heartbeat took 4.5s
W180117 17:46:51.103536 179 storage/node_liveness.go:365  [n4,hb] failed node liveness heartbeat: context deadline exceeded 
```

Here are the logs from one of the nodes still alive:
``` I180116 06:55:12.011036 114 storage/replicate_queue.go:230  [replicate,n1,s1,r59/1:/Table/52/1/"\x{be\x9…-e3\x0…}] snapshot failed: context deadline exceeded
E180116 06:55:12.011120 114 storage/queue.go:668  [replicate,n1,s1,r59/1:/Table/52/1/"\x{be\x9…-e3\x0…}] failed to replicate after 5 retries
I180116 06:55:12.013834 114 storage/replica_raftstorage.go:540  [replicate,n1,s1,r146/4:/Table/54/2{-/2017-0…}] generated preemptive snapshot c982b665 at index 18452
I180116 06:55:13.430883 124 server/status/runtime.go:219  [n1] runtime stats: 813 MiB RSS, 136 goroutines, 231 MiB/170 MiB/468 MiB GO alloc/idle/total, 335 MiB/408 MiB CGO al
loc/total, 589.28cgo/sec, 0.81/0.15 %(u/s)time, 0.00 %gc (0x)
I180116 06:55:23.435773 124 server/status/runtime.go:219  [n1] runtime stats: 809 MiB RSS, 136 goroutines, 294 MiB/116 MiB/468 MiB GO alloc/idle/total, 341 MiB/411 MiB CGO alloc/total, 635.81cgo/sec, 0.89/0.09 %(u/s)time, 0.00 %gc (0x)
W180116 06:55:32.298863 95 vendor/google.golang.org/grpc/clientconn.go:934  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 165.227.190.30:26257: getsockopt: connection refused"; Reconnecting to {165.227.190.30:26257 0  <nil>}
I180116 06:55:33.433143 124 server/status/runtime.go:219  [n1] runtime stats: 809 MiB RSS, 136 goroutines, 213 MiB/186 MiB/468 MiB GO alloc/idle/total, 342 MiB/411 MiB CGO alloc/total, 636.90cgo/sec, 0.86/0.12 %(u/s)time, 0.00 %gc (1x)
I180116 06:55:43.430943 124 server/status/runtime.go:219  [n1] runtime stats: 810 MiB RSS, 136 goroutines, 277 MiB/132 MiB/468 MiB GO alloc/idle/total, 343 MiB/412 MiB CGO alloc/total, 590.51cgo/sec, 0.70/0.15 %(u/s)time, 0.00 %gc (0x)
I180116 06:55:53.408375 122 gossip/gossip.go:487  [n1] gossip status (ok, 4 nodes)
gossip client (1/3 cur/max conns)
  3: 165.227.110.69:26257 (27h25m1s: infos 15268/462479 sent/received, bytes 2585339B/55766926B sent/received)
gossip server (0/3 cur/max conns, infos 1039742/636311 sent/received, bytes 136960865B/99592429B sent/received)
I180116 06:55:53.439832 124 server/status/runtime.go:219  [n1] runtime stats: 808 MiB RSS, 136 goroutines, 197 MiB/196 MiB/468 MiB GO alloc/idle/total, 342 MiB/411 MiB CGO alloc/total, 634.25cgo/sec, 0.77/0.12 %(u/s)time, 0.00 %gc (1x)
I180116 06:56:03.429598 124 server/status/runtime.go:219  [n1] runtime stats: 809 MiB RSS, 136 goroutines, 268 MiB/138 MiB/468 MiB GO alloc/idle/total, 343 MiB/413 MiB CGO alloc/total, 606.20cgo/sec, 0.72/0.15 %(u/s)time, 0.00 %gc (0x)
I180116 06:56:13.424864 124 server/status/runtime.go:219  [n1] runtime stats: 809 MiB RSS, 135 goroutines, 180 MiB/210 MiB/468 MiB GO alloc/idle/total, 343 MiB/412 MiB CGO alloc/total, 576.38cgo/sec, 0.72/0.16 %(u/s)time, 0.00 %gc (1x)
E180116 06:56:19.855247 19525789 server.go:2846  http: TLS handshake error from 118.106.34.105:4301: tls: first record does not look like a TLS handshake
I180116 06:56:23.431680 124 server/status/runtime.go:219  [n1] runtime stats: 818 MiB RSS, 136 goroutines, 244 MiB/159 MiB/468 MiB GO alloc/idle/total, 344 MiB/413 MiB CGO alloc/total, 637.58cgo/sec, 0.81/0.14 %(u/s)time, 0.00 %gc (0x)
W180116 06:56:27.060960 114 storage/raft_transport.go:649  [replicate,n1,s1,r146/4:/Table/54/2{-/2017-0…}] failed to close snapshot stream: rpc error: code = DeadlineExceeded desc = context deadline exceeded ```

**Comment by benesch** [comment]
@dianasaur323 can we look at this together?

**Comment by dianasaur323** [comment]
@benesch @nvanbenschoten is taking a look - unless you also wanted to take a look?

**Comment by nvanbenschoten** [comment]
What I'm seeing right now is that a number of empty ranges are having a hard time up-replicating because snapshots sent by `changeReplicas` keep timing out. I can see the following in a goroutine dump:
```
goroutine 114 [syscall, 1 minutes, locked to thread]:
github.com/cockroachdb/cockroach/pkg/storage/engine._Cfunc_DBIterSeek(0x7f940c221058, 0xc4267118b0, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/cockroachdb/cockroach/pkg/storage/engine/_obj/_cgo_gotypes.go:517 +0x6f
github.com/cockroachdb/cockroach/pkg/storage/engine.(*rocksDBIterator).Seek.func2(0x7f940c221058, 0xc4267118b0, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:1807 +0xc2
github.com/cockroachdb/cockroach/pkg/storage/engine.(*rocksDBIterator).Seek(0xc4292f7b60, 0xc4267118b0, 0x2, 0x8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:1807 +0x1b8
github.com/cockroachdb/cockroach/pkg/storage/rditer.(*ReplicaDataIterator).advance(0xc428510b70)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rditer/replica_data_iter.go:121 +0x59
github.com/cockroachdb/cockroach/pkg/storage/rditer.(*ReplicaDataIterator).Next(0xc428510b70)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rditer/replica_data_iter.go:108 +0x42
github.com/cockroachdb/cockroach/pkg/storage.sendSnapshot(0x34163c0, 0xc429bd7500, 0xc4200c7900, 0x7f9413704488, 0xc43fb22a30, 0x33f9cc0, 0xc42021fe00, 0x4818, 0x4804, 0xc429bd76e0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3537 +0x84a
github.com/cockroachdb/cockroach/pkg/storage.(*RaftTransport).SendSnapshot(0xc42079d950, 0x34163c0, 0xc429bd7500, 0xc42021fe00, 0x4818, 0x4804, 0xc429bd76e0, 0xc426e7e370, 0xc4267119e0, 0xc4267119a0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/raft_transport.go:652 +0x279
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).sendSnapshot(0xc420ab0a80, 0x34163c0, 0xc429bd7500, 0x200000002, 0x0, 0x22e0815, 0xa, 0x1, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_command.go:1867 +0x43a
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).changeReplicas(0xc420ab0a80, 0x34163c0, 0xc429bd7500, 0x200000000, 0x2, 0xc42c377260, 0xc400000001, 0x22f359a, 0x16, 0xc4207a9200, ...)
```

I'm not sure exactly why iteration over an empty range would take so long, but I expect that it's related to a large buildup of RocksDB tombstones. According to @dianasaur323, the cluster has gone through a series of `DROP DATABASE` and `RESTORE` cycles, which fits the bill for a potential slowdown in RocksDB. I'm going to play around with `cockroach debug compact`, `cockroach debug rocksdb`, and `cockroach debug estimate-gc` to see if anything pops out.

**Comment by nvanbenschoten** [comment]
Also note that if these `DROP DATABASE` and `RESTORE` cycles are to blame, then this issue should be significantly alleviated in the future by https://github.com/cockroachdb/cockroach/pull/20601.

**Comment by knz** [comment]
I think this has been addressed in the last batch of 2.1 updates. Closing.
